PlayTools/Controls/Frontend/ControlMode.swift

```diff
@@ -34,13 +34,13 @@ public class ControlMode: Equatable {
         let centre = NotificationCenter.default
         let main = OperationQueue.main
         if PlaySettings.shared.noKMOnInput {
-            centre.addObserver(forName: UIApplication.keyboardDidHideNotification, object: nil, queue: main) { _ in
-                ModeAutomaton.onKeyboardHide()
-                Toucher.writeLog(logMessage: "virtual keyboard did hide")
+            centre.addObserver(forName: UITextField.textDidEndEditingNotification, object: nil, queue: main) { _ in
```

Comment:
So instead of watching keyboardDidHide or show to trigger automaton methods you change it to watch any text inputs to be edited. Wouldn't this increase the amount of observed actions? whereas keyboard its only one observer?

Reply:
I read [the document](https://developer.apple.com/documentation/uikit/uitextfield/1619616-textdidbegineditingnotification), it's not watching any text input being edited, which would be another notification `textDidChangeNotification`, while it's `begin` and `end` ones here. So I believe actions amount is not increased.